ledger/src/blockstore.rs

```diff
@@ -2972,12 +2972,13 @@ impl Blockstore {
 
     // Returns slots connecting to any element of the list `slots`.
     pub fn get_slots_since(&self, slots: &[u64]) -> Result<HashMap<u64, Vec<u64>>> {
-        // Return error if there was a database error during lookup of any of the
-        // slot indexes
-        let slot_metas: Result<Vec<Option<SlotMeta>>> =
-            slots.iter().map(|slot| self.meta(*slot)).collect();
+        let slot_metas: Vec<_> = self
+            .meta_cf
+            .multi_get(slots.to_vec())
+            .into_iter()
+            .map(|meta| meta.unwrap())
```

Comment:
@steviez: you probably know better than me on this.  Is there a nicer / rust way to handle this unwrap?

Reply:
Yeah, I think we'd still prefer to check Result's with ? instead of .unwrap() (recalling that a bad unwrap() will just panic).

Doing .into_iter().collect() allows you to transform a Vec<Result<...>> into Result<Vec<...>>, so something like this:
```rust
        let slot_metas: Result<Vec<Option<SlotMeta>>> = self
            .meta_cf
            .multi_get(slots.to_vec())
            .into_iter()
            .collect();
        let slot_metas = slot_metas?;
```

Unfortunately you can't do a 1-liner with the ? immediately after .collect(), the compiler complains about not being able to infer type

Reply:
Actually, I guess you can 1-liner it by moving the type declaration into `.collect()`:
```rust
        let slot_metas = self
            .meta_cf
            .multi_get(slots.to_vec())
            .into_iter()
            .collect::<Result<Vec<Option<SlotMeta>>>>()?;
```
Can't say I have a preference for one vs. the other, I think our codebase uses the style of my first comment more than the style of this comment.

Reply:
```
        let slot_metas: Result<Vec<Option<SlotMeta>>> = self
            .meta_cf
            .multi_get(slots.to_vec())
            .into_iter()
            .collect();
        let slot_metas = slot_metas?;
```        
Interesting!  I `multi_get()` returns ` Vec<Result<Option<DBPinnableSlice>>>`, which makes me think that I need to stick with its original type.  But when we explicitly says we want `Result<Vec<Option<SlotMeta>>>` as the return type of `collect()`, then it just magically works.

Do you happen to know how rust converts `Vec<Result<Option<DBPinnableSlice>>>` to `Result<Vec<Option<SlotMeta>>>`?

Reply:
I believe you're confusing the two `multi_get()` calls:
```rust
Rocks::multi_get_cf(...) -> Vec<Result<Option<DBPinnableSlice>>>

LedgerColumn<C>::multi_get(...) -> Vec<Result<Option<C::Type>>>
```
And recall that the latter calls `bincode::deserialize()` to map `DBPinnableSlice` to `C::Type`; this function is able to deduce the type that `bincode::deserialize()` yields based on overall function return type (`C::Type` on this column is `SlotMeta`).

So, not as magical as you thought but still pretty nifty 😉

Reply:
I still don't fully understand, but we are changing from `Vec<Result<...>> to `Result<Vec<...>>`, right?

Reply:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect